### DISCUSSION TITLE
Unable to fetch result from OAuth login

### DISCUSSION THREAD

**Discussion by suyash515** [Web3Auth/Web3Auth#515] [Help]
![openlogin-error](https://user-images.githubusercontent.com/5890499/185937341-a047e1e7-89b4-48f4-9381-16be23fb3261.png)

I am getting the following issue on my end and these are the steps to reproduce the error:

1. Create a Custom Auth (Login Provider: Custom)
2. JWT Verifier ID: Sub
3. Set up the JWK Endpoint
4. JWT Validation: iss field set up
5. Set up a Plug and Play Project and using EVM based chain
6. Using the following code for the integration:

```
const web3authCore = new Web3AuthCore({
			chainConfig: { // this is ethereum chain config, change if other chain(Solana, Polygon)
				chainNamespace: CHAIN_NAMESPACES.EIP155,
				chainId: "0x1",
				rpcTarget: "mainnet_infura_url",
				blockExplorer: "https://etherscan.io/",
				ticker: "ETH",
				tickerName: "Ethereum"
			},
		});
		const adapter = new OpenloginAdapter({ adapterSettings: {
			network: "testnet",
			clientId: "client_id_from_web3auth_from_plugnplay",
			uxMode: "popup", // other option: popup
			loginConfig: {
				jwt: {
					name: "any name",
					verifier: "test",
					typeOfLogin: "jwt",
					clientId: "client_id_from_our_custom_login_provider",
				},
			},
		}});
		web3authCore.configureAdapter(adapter);
		await web3authCore.init();

		// call below code when user clicks on login button
		// it will use custom login with openlogin's authentication
		await web3authCore.connectTo(adapter.name, { 
			loginProvider: "jwt",
			extraLoginOptions: {
				domain: "redirect_uri_to_custom_login_provider",
				verifierIdField: "sub", 
				response_type: "token",
				scope:"email", // e.g. email openid
			}
		});
```

_Originally posted by @suyash515 in https://github.com/Web3Auth/Web3Auth/discussions/446#discussioncomment-3447463_

**Comment by rinzler99** [comment]
pls share console errors, if any

**Comment by suyash515** [comment]
![openlogin-console-error](https://user-images.githubusercontent.com/5890499/185980656-0e1bb307-b307-428f-9a60-8a94ff60b0d3.png)

@rinzler99 console errors.

**Comment by shahbaz17** [comment]
Hey @suyash515 

Which login provider are you using? This looks like the authentication didn't happen.

If you're, you are using a custom authentication provider(that are not natively supported on the verifier dashboard) that returns `idToken` after successful authentication. Then, simply use the below code to log in with Web3Auth:

```js
import { Web3AuthCore } from "@web3auth/core";
import { WALLET_ADAPTERS, CHAIN_NAMESPACES } from "@web3auth/base";

const web3auth = new Web3AuthCore({
  chainConfig: {
    chainNamespace: CHAIN_NAMESPACES.EIP155, // SOLANA, OTHER
    chainId: "0x1",
    rpcTarget: "https://rpc.ankr.com/eth",
    displayName: "Ethereum Mainnet",
    blockExplorer: "https://etherscan.io",
    ticker: "ETH",
    tickerName: "Ethereum",
  },
});

const openloginAdapter = new OpenloginAdapter({
  adapterSettings: {
    clientId: "YOUR-WEB3AUTH-CLIENT-ID-FROM-PLUG-AND-PLAY",
    network: "testnet",
    uxMode: "redirect", // also support popup
    loginConfig: {
      jwt: {
        name: "Name of your choice",
        verifier: "YOUR-VERIFIER-NAME-ON-WEB3AUTH-DASHBOARD-FROM-CUSTOM-AUTH",
        typeOfLogin: "jwt",
        clientId: "YOUR-WEB3AUTH-CLIENT-ID-FROM-PLUG-AND-PLAY",
      },
    },
  },
});

web3auth.configureAdapter(openloginAdapter);

await web3auth.init();

await web3auth.connectTo(WALLET_ADAPTERS.OPENLOGIN, {
  loginProvider: "jwt",
  extraLoginOptions: {
    id_token: idToken,
    verifierIdField: "sub", // same as your JWT Verifier ID
    domain: "https://YOUR-APPLICATION-DOMAIN" || "http://localhost:3000",
  },
});
```

Also, Please follow each step mentioned https://web3auth.io/docs/custom-authentication/byo-jwt-providers#set-up-custom-jwt-verifier for setting a Verifier.

**Comment by suyash515** [comment]
> Hey @suyash515
> 
> Which login provider are you using? This looks like the authentication didn't happen.
> 
> If you're, you are using a custom authentication provider(that are not natively supported on the verifier dashboard) that returns `idToken` after successful authentication. Then, simply use the below code to log in with Web3Auth:
> 
> ```js
> import { Web3AuthCore } from "@web3auth/core";
> import { WALLET_ADAPTERS, CHAIN_NAMESPACES } from "@web3auth/base";
> 
> const web3auth = new Web3AuthCore({
>   chainConfig: {
>     chainNamespace: CHAIN_NAMESPACES.EIP155, // SOLANA, OTHER
>     chainId: "0x1",
>     rpcTarget: "https://rpc.ankr.com/eth",
>     displayName: "Ethereum Mainnet",
>     blockExplorer: "https://etherscan.io",
>     ticker: "ETH",
>     tickerName: "Ethereum",
>   },
> });
> 
> const openloginAdapter = new OpenloginAdapter({
>   adapterSettings: {
>     clientId: "YOUR-WEB3AUTH-CLIENT-ID-FROM-PLUG-AND-PLAY",
>     network: "testnet",
>     uxMode: "redirect", // also support popup
>     loginConfig: {
>       jwt: {
>         name: "Name of your choice",
>         verifier: "YOUR-VERIFIER-NAME-ON-WEB3AUTH-DASHBOARD-FROM-CUSTOM-AUTH",
>         typeOfLogin: "jwt",
>         clientId: "YOUR-WEB3AUTH-CLIENT-ID-FROM-PLUG-AND-PLAY",
>       },
>     },
>   },
> });
> 
> web3auth.configureAdapter(openloginAdapter);
> 
> await web3auth.init();
> 
> await web3auth.connectTo(WALLET_ADAPTERS.OPENLOGIN, {
>   loginProvider: "jwt",
>   extraLoginOptions: {
>     id_token: idToken,
>     verifierIdField: "sub", // same as your JWT Verifier ID
>     domain: "https://YOUR-APPLICATION-DOMAIN" || "http://localhost:3000",
>   },
> });
> ```
> 
> Also, Please follow each step mentioned https://web3auth.io/docs/custom-authentication/byo-jwt-providers#set-up-custom-jwt-verifier for setting a Verifier.

@shahbaz17 Thanks so much for the reply. 
So, if I understand well, the custom login provider should return an additional parameter called idToken, right? What should the parameter contain?

The reason is that we have control over the custom login provider and can have any required values returned based on your requirements.

**Comment by shahbaz17** [comment]
Yes, that's correct. That should return an `idToken` in JWT format.

* JWT token header must contain the `kid` field.
* JWT token payload data must contain `iat` field.
* JWT token payload data must contain `iss` field. The value of this field will be used in JWT Validation field of Custom Auth Verifier on Web3Auth Dashboard..
* JWT token payload data must contain `aud` field. The value of this field will be used in JWT Validation field of Custom Auth Verifier on Web3Auth Dashboard.
* JWKS used to sign the token, should be exposed to Web3Auth. This endpoint URL to be used in the JWKS Endpoint field of Custom Auth Verifier on Web3Auth Dashboard.

Check https://web3auth.io/docs/custom-authentication/byo-jwt-providers#using-jsonwebtoken for more details on how to generate an idToken.

Using jose, it would look something like this:
```js
import * as jose from "jose";

import fs from "fs";
// openssl genrsa -out privateKey.pem 512
var privateKey = fs.readFileSync("privateKey.pem");
// openssl rsa -in privateKey.pem -pubout -out publicKey.pem
var publicKey = fs.readFileSync("publicKey.pem");
// https://my-authz-server/.well-known/jwks.json -> to be used in Custom Authentication as JWK Endpoint.

// Signing the JWT
const jwt = await new jose.SignJWT({ "urn:example:claim": true })
  .setProtectedHeader({ alg: "RS256", kid: "1bb9605c36e{your_kid}69386830202b2d" })
  .setIssuedAt()
  .setIssuer("https://my-authz-server")
  .setAudience("urn:my-resource-server")
  .setExpirationTime("2h")
  .sign(privateKey);
 ```

**Comment by suyash515** [comment]
@shahbaz17 Thank you. Let me try this and revert back.

**Comment by shahbaz17** [comment]
Hey @suyash515 

Have you tried this?

**Comment by suyash515** [comment]
@shahbaz17 We are still working on this. I will have an update on this later today.

**Comment by suyash515** [comment]
@shahbaz17 we have tried the above solution but getting the following error:
![image (2)](https://user-images.githubusercontent.com/5890499/187412347-a2a90567-ec88-4a52-a868-6c278c84b9fb.png)

This is the code that I am using:

            `payload = {
                sub: "access_token_that_we_generated"
            }

            const [key] = keyStore.all({
                use: "sig"
            });

            const opt = {
                compact: true,
                jwk: key,
                fields: {
                    typ: "jwt",
                    iss: "https://lootverse.com/",
                    kid: keyStore.toJSON().keys[0].kid,
                    aud: "versa",
                    algorithm: "RS256"
                }
            };

            token = await jose.JWS.createSign(opt, key).update(JSON.stringify(payload)).final();`

Basically, we are redirecting back to https://beta.openlogin.com/auth?idToken=token

**Reply by shahbaz17** [reply]
Can you confirm which version of web3auth you are using?

Try updating to the latest. 

> Note: there's a breaking change, and you would need to pass clientId at different places.

**Reply by shahbaz17** [reply]
Can you share a sample idToken here? That you're generating.

**Reply by suyash515** [reply]
We are currently using the @web3auth/base v1.2.0 and @web3auth/web3auth v1.2.0

This is the idToken that we are using:
`eyJ0eXAiOiJKV1QiLCJpc3MiOiJodHRwczovL2xvb3R2ZXJzZS5jb20iLCJraWQiOiJMQ2NrR24xWU8xNTJpcXJCd1cxUjBKLTdqd1MydXNRSmc0UXZkb094ZXVVIiwiYXVkIjoidmVyc2EiLCJhbGdvcml0aG0iOiJSUzI1NiIsImFsZyI6IlJTMjU2In0.eyJzdWIiOiJOM3JENDVpV3VnclZxdjh0Q0lIVGhQUEN5eXRhb1I3cHVkSHpBSGxOcXBIdm55UjFZRUhRM1pZb1FVMGlaY2pQMmdRbWVvcTJyUGgydnAwNkNpeDJjMjFKN3NNN245OTkzbDNFIn0.BjHq65CyF4Zf1gPX2wEU4SIWJB7BZoU19BQNyLZ_xK96IqaSHsfmHNWRjTETXEU9V_XOXbVVvqKj6x6oZdIEyODFxHwC7Hgr6EeshG0t9wxJP_2N45yh3sX8h4iBXfTLojPFd3Dq_2CzVxoq8iXNuRZID2mdPvPvHbqYjtJrL6uPfHdcvkgeZkrPBa0Oyci59F-y4FrU1epVlND68WypvnsBAZdjv2hkdzZHJqnlCJdnBpn-81tfg_dKq6e1VvfpcKqnDEBqbRbg0o8SYpzWpKTMXxjUugqRFr9Ne3nqngqi-5tElYd1JBa5vE8EB_X3MZo92gz2y7j_VHIeuAptoA`

**Reply by suyash515** [reply]
@shahbaz17 based on your comment above, does it mean that in the return url: https://beta.openlogin.com/auth, I also need to pass the clientId as a GET parameter?

**Reply by shahbaz17** [reply]
<img width="616" alt="Screenshot 2022-08-30 at 5 00 51 PM" src="https://user-images.githubusercontent.com/6962565/187425793-247c6537-fc7e-4809-972e-61e53f3ffc2a.png">

Just decoded your JWT Token on https://jwt.io, and it has wrong data at places.

JWT Header should only have:
```json
{
  "typ": "JWT",
  "kid": "LCckGn1YO152iqrBwW1R0J-7jwS2usQJg4QvdoOxeuU",
  "alg": "RS256"
}
```

JWT Payload data should have the following fields:
```json
{
  "iat": 124567238,
  "iss": "https://lootverse.com",
  "aud": "versa",
  "sub": "N3rD45iWugrVqv8tCIHThPPCyytaoR7pudHzAHlNqpHvnyR1YEHQ3ZYoQU0iZcjP2gQmeoq2rPh2vp06Cix2c21J7sM7n9993l3E",
  "exp": 124568238
}
```

And make sure to set up your `Custom Auth Verifier` on Web3Auth Dashboard correctly, using https://web3auth.io/docs/custom-authentication/byo-jwt-providers#set-up-custom-jwt-verifier

And JWT Validation should match the data present in your JWT Token from here.

**Reply by shahbaz17** [reply]
And Expose your JWKS endpoint to a URL and input it under the JWKS Endpoint field of the Verifier.

**Reply by shahbaz17** [reply]
Also, JWT token payload data must contain `iat` field.

**Reply by suyash515** [reply]
@shahbaz17 thanks for all the information.

So, I updated my code and these are the results:
The page is being redirected to the following url: https://beta.openlogin.com/auth?idToken=eyJ0eXAiOiJKV1QiLCJraWQiOiI3MXpOR1ByUWRFUml6dG9rdEJzOGtpWG9FbUdsM2QzWTJDTG5hS3hrZWNBIiwiYWxnIjoiUlMyNTYifQ.eyJzdWIiOiJnSlQ5aDFNQkJsbEVWcFpIdVRUdjlSMXh2dk91QjZFRUhGMHRLdlROTVpYVmF2R3JTRHIyQ2xycDZqOUdPQmdJYjBrSUdtZzNYWVJFVWNreDBoY2tmN2lMbUw4M1hiTFJWRTZQIiwiaWF0IjoxNjYxOTM4NTk0LCJleHAiOjE2NjIwMjQ5OTQsImlzcyI6ImJldGEubG9vdHZlcnNlLmNvbSIsImF1ZCI6InZlcnNhIn0.CP5BpIOX1N7BMTOyBFIwE8VP1pvGmvYLxUcRBDhS0gQ1uLKHgX17zsif7OGwLlDqB1Hmu6HQKo6xTXVJbv9eg_hAcBi4tjoUeMcYeeU4IJ2kkJjHAw6rbSn2r4459HS3PKz01gE3nxwv3haAUQn-QDZLTJv1_WnaIT3EKSuZKq2E28kFjsaHIIlc4VFWIuwBI5gRcsBimkXtLWEoh_ZOsbggTpQciqO2YPHEQ-d8r8BzlBEvjUNOsw0Xcto92DtqwetmUstxkv7DiIuIbdzCqIGQaP9nXutaFxD46c5xlMk5VDfuHJ9XDqK43rVcAWeL5ZrzWHEGApE1SJrGRy21OQ

Additional information:
1. The jwt includes the payload that you mentioned
2. I have checked the jwt signature and it's been verifed on the jwt.io site (the link to get the jwks is here: https://public-api-staging.lootnft.io/api/v1/oauth/jwks.js)
3. The jwks link has already been added to the JWK Endpoint in the Web3Auth Custom Verifier screen

**Reply by shahbaz17** [reply]
And you're using this code to login, right?

```js
import { WALLET_ADAPTERS, CHAIN_NAMESPACES } from "@web3auth/base";

await web3auth.connectTo(WALLET_ADAPTERS.OPENLOGIN, {
  loginProvider: "jwt",
  extraLoginOptions: {
    id_token: idToken,
    verifierIdField: "sub", // same as your JWT Verifier ID
    domain: "https://YOUR-APPLICATION-DOMAIN" || "http://localhost:3000",
  },
});
```

**Reply by suyash515** [reply]
I have the following code for login:

```
await web3authCore.connectTo(adapter.name, { 
	loginProvider: "jwt",
	extraLoginOptions: {
		domain: "https://YOUR-APPLICATION-DOMAIN",
		verifierIdField: "sub", 
		response_type: "token",
		scope:"email",
	}
});
```
Now the idToken is generated in the custom verifier, which is a separate standalone system. The question is really: how would I retrieve the idToken in the login code?

Please note that to log in, I'm using whatever is generated in the https://dashboard.web3auth.io/home/customauth page.

**Reply by shahbaz17** [reply]
Please use the code I share just above. The code you're using is for getting the idToken with a scope of email. This is different and has different purpose.

**Reply by shahbaz17** [reply]
Please read https://web3auth.io/docs/guides/cognito#additional-reading-setup-custom-authentication-using-aws-cognito-authorization-code-flow or https://web3auth.io/docs/guides/auth0#additional-reading-setup-custom-authentication-using-auth0-rwa to understand how and why we need this way.

**Reply by shahbaz17** [reply]
![Screenshot 2022-09-02 at 12 18 03 PM](https://user-images.githubusercontent.com/6962565/188075764-1aceddc5-6d28-4369-858d-03e2e76133aa.png)

This is what we want to achieve here. Since you already have a way to generate idToken, you're passing that token to web3auth to help you with the private-public key pair.

**Reply by suyash515** [reply]
@shahbaz17 we changed our implementation and we went one step further in the process but now encountering the one of the following errors:
1. unable to detect login share from the Auth Network. This may be due to slow internet connection...
2. there seems to be some bug in the code. Please contact support to fix this. You system time is incorrect. Please correct your system time and login again.

However, in our dashboard in Web3Auth, we can see that we have 2 connected users.


![incorrect-system-time](https://user-images.githubusercontent.com/5890499/189640255-cd716a7b-44ef-4426-b392-51e012ba2387.png)
![detect-login-share](https://user-images.githubusercontent.com/5890499/189640265-23e2eb4a-d1df-4021-9b20-482fd605081f.png)
![constructing-key](https://user-images.githubusercontent.com/5890499/189640280-5ceb3b6b-b5a5-48da-9a22-9aae30d9a182.png)

**Reply by suyash515** [reply]
@shahbaz17 Hey there.

Do you have any idea what the above error could be related to? Thank you.

**Reply by shahbaz17** [reply]
Hey @suyash515 

Is there a way to see your code?

Meanwhile, can you [schedule a call with our team](https://calendly.com/web3auth/meeting-with-web3auth), and we can go together to debug your code?

Add this discussion link while scheduling the call.

**Comment by ChaviJ28** [comment]
Hi @shahbaz17 
The following code is being used to login : 
```
    const web3authCore = new Web3AuthCore({
      chainConfig: { // this is ethereum chain config, change if other chain(Solana, Polygon)
        chainNamespace: CHAIN_NAMESPACES.EIP155,
        chainId: "0x1",
        rpcTarget: "https://mainnet.infura.io/v3/776218ac4734478c90191dde8cae483c",
        blockExplorer: "https://etherscan.io/",
        ticker: "ETH",
        tickerName: "Ethereum"
      },
    });

    const adapter = new OpenloginAdapter({
      adapterSettings: {
        network: "testnet",
        clientId: CLIENT_ID_WEB3AUTH,
        uxMode: "popup", // other option: popup, redirect
        loginConfig: {
          jwt: {
            name: "any name",
            verifier: "codevigor-lootverse-testnet",
            typeOfLogin: "jwt",
            clientId: LOOTVERSE_CLIENT_ID,
          },
        },
      }
    });
    web3authCore.configureAdapter(adapter);
    await web3authCore.init();

    // call below code when user clicks on login button
    // it will use custom login with openlogin's authentication
    await web3authCore.connectTo(adapter.name, {
      loginProvider: "jwt",
      extraLoginOptions: {
        domain: "https://public-api-staging.lootnft.io/",
        id_token: idToken,
        verifierIdField: "sub",
      }
    });
```

idToken is in JWT format containing the following data: 
```
{
  "sub": LOOTVERSE_ACCESS_TOKEN,
  "iat": 1663593984,
  "exp": 1663680384,
  "iss": "beta.lootverse.com",
  "aud": "versa"
}
```

**Reply by suyash515** [reply]
@shahbaz17 thanks for your reply. Please see above the code that we are using. I also booked a call with the team tomorrow. Thank you for your help.

**Comment by shahbaz17** [comment]
This is resolved now.

**Comment by himanshuchawla009** [comment]
is your issue resolved, If not , Can you also attach console logs along with the screenshots that you posted above. Also can you check if your third party cookies are blocked while using solana wallet.
